File: localmind-sdk-js/src/conversations.ts
import type { LocalMindClient } from './client';
import type { Conversation, ConversationSummary } from './types';

/**
 * Client for conversation management operations.
 *
 * @example
 * ```typescript
 * const convos = client.conversations;
 * const all = await convos.list();
 * const full = await convos.getById("abc-123");
 * const pdf = await convos.export("abc-123", "pdf");
 * await convos.delete("abc-123");
 * ```
 */
export class ConversationClient {
  constructor(private readonly client: LocalMindClient) {}

  /**
   * List all conversations, optionally filtered by tag.
   */
  async list(tag?: string): Promise<ConversationSummary[]> {
    const params = tag ? `?tag=${encodeURIComponent(tag)}` : '';
    return this.client._get<ConversationSummary[]>(`/conversations${params}`);
  }

  /**
   * Get a conversation with full message history.
   */
  async getById(id: string): Promise<Conversation> {
    return this.client._get<Conversation>(`/conversations/${id}`);
  }

  /**
   * Rename a conversation.
   */
  async rename(id: string, title: string): Promise<Conversation> {
    return this.client._patch<Conversation>(`/conversations/${id}`, { title });
  }

  /**
   * Update the system prompt for a conversation.
   */
  async updateSystemPrompt(id: string, systemPrompt: string): Promise<Conversation> {
    return this.client._patch<Conversation>(`/conversations/${id}/system-prompt`, { systemPrompt });
  }

  /**
   * Add a tag to a conversation.
   */
  async addTag(id: string, tag: string): Promise<Conversation> {
    return this.client._post<Conversation>(`/conversations/${id}/tags`, { tag });
  }

  /**
   * Remove a tag from a conversation.
   */
  async removeTag(id: string, tag: string): Promise<Conversation> {
    return this.client._deleteWithResponse<Conversation>(`/conversations/${id}/tags/${encodeURIComponent(tag)}`);
  }

  /**
   * Export a conversation in the specified format.
   *
   * @param id - Conversation UUID.
   * @param format - Export format: "json", "md", "markdown", or "pdf".
   * @returns Raw bytes of the exported file as an ArrayBuffer.
   */
  async export(id: string, format: string = 'json'): Promise<ArrayBuffer> {
    return this.client._getBytes(`/conversations/${id}/export?format=${format}`);
  }

  /**
   * Delete a conversation.
   */
  async delete(id: string): Promise<void> {
    await this.client._delete(`/conversations/${id}`);
  }
}

File: localmind-sdk-python/localmind/conversations.py
"""Conversation management client for LocalMind."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Dict, List, Optional

if TYPE_CHECKING:
    from .client import LocalMindClient


class ConversationClient:
    """Client for conversation management operations.

    Example::

        convos = client.conversations
        all_convos = convos.list()
        full = convos.get("abc-123")
        exported = convos.export("abc-123", format="pdf")
        convos.delete("abc-123")
    """

    def __init__(self, client: "LocalMindClient"):
        self._client = client

    def list(self, tag: Optional[str] = None) -> List[Dict[str, Any]]:
        """List all conversations, optionally filtered by tag.

        Args:
            tag: Optional tag to filter conversations.

        Returns:
            List of conversation summary objects.
        """
        params = {}
        if tag:
            params["tag"] = tag
        return self._client._get("/conversations", params=params)

    def get(self, conv_id: str) -> Dict[str, Any]:
        """Get a conversation with full message history.

        Args:
            conv_id: Conversation UUID.

        Returns:
            Full conversation object including messages.
        """
        return self._client._get(f"/conversations/{conv_id}")

    def rename(self, conv_id: str, title: str) -> Dict[str, Any]:
        """Rename a conversation.

        Args:
            conv_id: Conversation UUID.
            title: New title.

        Returns:
            Updated conversation object.
        """
        return self._client._patch(f"/conversations/{conv_id}", json={"title": title})

    def update_system_prompt(self, conv_id: str, system_prompt: str) -> Dict[str, Any]:
        """Update the system prompt for a conversation.

        Args:
            conv_id: Conversation UUID.
            system_prompt: New system prompt.

        Returns:
            Updated conversation object.
        """
        return self._client._patch(
            f"/conversations/{conv_id}/system-prompt",
            json={"systemPrompt": system_prompt},
        )

    def add_tag(self, conv_id: str, tag: str) -> Dict[str, Any]:
        """Add a tag to a conversation.

        Args:
            conv_id: Conversation UUID.
            tag: Tag string.

        Returns:
            Updated conversation object.
        """
        return self._client._post(f"/conversations/{conv_id}/tags", json={"tag": tag})

    def remove_tag(self, conv_id: str, tag: str) -> Dict[str, Any]:
        """Remove a tag from a conversation.

        Args:
            conv_id: Conversation UUID.
            tag: Tag to remove.

        Returns:
            Updated conversation object.
        """
        return self._client._delete_with_response(f"/conversations/{conv_id}/tags/{tag}")

    def export(self, conv_id: str, format: str = "json") -> bytes:
        """Export a conversation in the specified format.

        Args:
            conv_id: Conversation UUID.
            format: Export format - "json", "md", "markdown", or "pdf".

        Returns:
            Raw bytes of the exported file.
        """
        return self._client._get_bytes(
            f"/conversations/{conv_id}/export", params={"format": format}
        )

    def delete(self, conv_id: str) -> None:
        """Delete a conversation.

        Args:
            conv_id: Conversation UUID.
        """
        self._client._delete(f"/conversations/{conv_id}")
